xmodule/capa/xqueue_submission.py

```diff
@@ -0,0 +1,104 @@
+"""
+This module provides an interface for submitting student responses
+to an external grading system through XQueue.
+"""
+
+import json
+import logging
+
+from xmodule.capa.errors import (
+    GetSubmissionParamsError,
+    JSONParsingError,
+    MissingKeyError,
+    ValidationError,
+    TypeErrorSubmission,
+    RuntimeErrorSubmission
+)
+
+log = logging.getLogger(__name__)
+
+
+class XQueueInterfaceSubmission:
+    """Interface to the external grading system."""
+
+    def __init__(self, block):
+        self.block = block
+
+    def _parse_json(self, data, name):
+        """Helper function to parse JSON safely."""
+        try:
+            return json.loads(data) if isinstance(data, str) else data
+        except json.JSONDecodeError as e:
+            raise JSONParsingError(name, str(e)) from e
+
+    def get_submission_params(self, header, payload):
+        """
+        Extracts student submission data from the given header and payload.
+        """
+        header = self._parse_json(header, "header")
+        payload = self._parse_json(payload, "payload")
+
+        queue_name = header.get('queue_name', 'default')
+
+        if not self.block:
+            raise GetSubmissionParamsError()
+
+        course_id = str(self.block.scope_ids.usage_id.context_key)
+        item_type = self.block.scope_ids.block_type
+        block_id = self.block.scope_ids.def_id.block_id
+        points_possible = self.block.max_score()
+
+        item_id = f"block-v1:{course_id.replace('course-v1:', '')}+type@{item_type}+block@{block_id}"
```

Comment:
Never build block usage keys by hand, because it will break in various edge cases that will be hard for you to test, e.g. CCX courses that use a different prefix.

In this case though, why do we need to do this? it looks like we're just reconstructing `self.block.usage_key`?

Reply:
If you do need to create a UsageKey manually, you can do so by grabbing a LearningContextKey (in this case a CourseKey) and calling `.make_usage_key()` on it with the appropriate parameters.